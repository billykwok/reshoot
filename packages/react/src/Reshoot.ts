import { forwardRef } from 'react';
import { css } from 'linaria';
import assign from 'object-assign';
import { INITIAL, FADING, LOADED } from './state';
import useLoadingState from './hooks/useLoadingState';
import useForwardableRef from './hooks/useForwardableRef';
import useImgCallback from './hooks/useImgCallback';
import useDownload from './hooks/useDownload';
import useIntersection from './hooks/useIntersection';
import cx from './utils/cx';
import createElement from './utils/createElement';
import IS_BROWSER from './utils/isBrowser';

import type { SyntheticEvent, RefObject } from 'react';
import type { State } from './state';

const asContainer = css`
  display: inline-block;
  position: relative;
  padding: 0;
  overflow: hidden;
  border: none;
  outline: none;
  text-decoration: none;
  color: transparent;
  background: currentColor;
  button& {
    cursor: pointer;
  }
  img {
    display: block;
    position: absolute;
    top: 0;
    width: 100%;
    height: 100%;
    object-fit: cover;
  }
  &::before {
    content: '';
    display: block;
    width: 100%;
    height: 0;
    padding-bottom: calc(var(--r, 0) * 100%);
  }
`;

const asPlaceholder = css`
  position: absolute;
  top: 0;
  width: 100%;
  height: 100%;
  background: currentColor;
  transform: scale3d(1.05, 1.05, 1);
  > img {
    filter: blur(0.5rem);
  }
`;

const asFadeout = css`
  animation: f 0.75s 1;
  @keyframes f {
    0% {
      transform: scale3d(1.05, 1.05, 1);
      opacity: 1;
    }
    100% {
      transform: none;
      opacity: 0;
    }
  }
`;

const asMessage = css`
  display: block;
  width: 100%;
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto,
    'Helvetica Neue', Arial, sans-serif;
  text-align: center;
  color: #fff;
  padding: 1rem 1.25rem;
  background: rgba(0, 0, 0, 0.5);
  border-radius: 0.25rem;
  max-width: 8rem;
  > svg {
    width: 4rem;
    height: 4rem;
    > path {
      stroke: #fff;
      stroke-width: 4;
      fill: none;
    }
  }
  > h6 {
    font-size: 1rem;
    margin: 0 0 0.25rem;
  }
  > p {
    font-size: 0.875rem;
    margin: 0;
  }
`;

const MESSAGES: [string, string, string] = [
  'Not autoloaded in slow network',
  'Browser is offline',
  'Fail to load',
];

type Props = Readonly<{
  className?: string;
  config: Readonly<{
    src: string;
    width: number;
    height: number;
    aspectRatio?: number;
    srcSet?: string;
    alt?: string;
    color?: string;
    placeholder?: string | false | null;
  }>;
  messages: Readonly<[string, string, string]>;
  imgProps: Readonly<Record<string, unknown>>;
  onLoad?: () => void;
  onError?: (event: Event | SyntheticEvent | string) => void;
  _s?: State;
  [key: string]: Readonly<unknown>;
}>;

const Reshoot = (
  {
    className,
    style,
    config: {
      src,
      width,
      height,
      aspectRatio,
      srcSet,
      alt,
      color,
      placeholder,
    },
    messages = MESSAGES,
    imgProps,
    onLoad,
    onError,
    _s = null,
    ...extraProps
  }: Props,
  ref: RefObject<HTMLElement>
) => {
  const [state, setState] = useLoadingState(_s, src);
  const _ref = useForwardableRef<HTMLElement>(ref);
  const callbacks = useImgCallback(_ref, src, state, setState, onLoad, onError);
  const download = useDownload(src, srcSet, callbacks);
  useIntersection(_ref, state, download);
  const dimensions = { width, height };
  const hasError = IS_BROWSER && state < INITIAL;

  return createElement(
    hasError ? 'button' : extraProps.href ? 'a' : 'div',
    assign(
      {
        ref: _ref,
        className: cx(asContainer, className),
        style: assign(
          { color, '--r': aspectRatio ? 1 / aspectRatio : height / width },
          style
        ),
      },
      extraProps,
      hasError && {
        onClick: () => {
          setState(() => INITIAL);
          download();
        },
      }
    ),
    createElement(
      'img',
      assign(
        { alt },
        (!IS_BROWSER || state > INITIAL) && { src, srcSet },
        dimensions,
        imgProps
      )
    ),
    IS_BROWSER &&
      placeholder &&
      state !== LOADED &&
      createElement(
        'div',
        {
          className: cx(asPlaceholder, state === FADING && asFadeout),
          onAnimationEnd: () => setState(() => LOADED),
        },
        createElement(
          'img',
          assign({ src: placeholder, alt: '', loading: 'lazy' }, dimensions)
        )
      ),
    hasError &&
      createElement(
        'div',
        { className: asMessage },
        createElement(
          'svg',
          { viewBox: '0 0 100 100' },
          createElement('path', {
            d: 'M79.4 56a30 30 0 1 1-1.7-17.5m2-19v20h-20',
          })
        ),
        createElement('h6', {}, messages[state as 0 | 1 | 2]),
        createElement('p', {}, 'Click to reload')
      )
  );
};

export default forwardRef(Reshoot);
