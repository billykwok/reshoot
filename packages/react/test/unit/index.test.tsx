import React from 'react';
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom/extend-expect';

import Reshoot from '../../src';

describe('react', () => {
  beforeAll(() => {
    (window as any).IntersectionObserver = class IntersectionObserver {
      observe() {
        return null;
      }

      unobserve() {
        return null;
      }
    };
  });

  test('parse valid input', () => {
    expect(
      renderer
        .create(<Reshoot src="image.jpg" alt="Test image" aspectRatio={80} />)
        .toJSON()
    ).toMatchSnapshot();
  });

  test('customized options for image link', () => {
    expect(
      renderer
        .create(
          <Reshoot
            src="image.jpg"
            alt="Test image"
            aspectRatio={80}
            blur={100}
            color="#fff"
            placeholder="placeholder.jpg"
            target="_blank"
            href="https://example.com"
            messages={{
              MANUAL: 'manual!',
              OFFLINE: 'offline!',
              ERROR: 'error!'
            }}
          />
        )
        .toJSON()
    ).toMatchSnapshot();
  });

  test('customized options for normal image', () => {
    expect(
      renderer
        .create(
          <Reshoot
            src="image.jpg"
            alt="Test image"
            aspectRatio={80}
            blur={100}
            color="#fff"
            messages={{
              MANUAL: 'manual!',
              OFFLINE: 'offline!',
              ERROR: 'error!'
            }}
          />
        )
        .toJSON()
    ).toMatchSnapshot();
  });
});
