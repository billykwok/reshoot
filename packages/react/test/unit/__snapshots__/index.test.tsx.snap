// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`react customized options for image link 1`] = `
<a
  className="a1y806g2"
  href="https://example.com"
  target="_blank"
>
  <div
    className="aund3wh"
    style={
      Object {
        "background": "#fff",
        "color": "#fff",
        "paddingBottom": "80%",
      }
    }
  />
  <img
    alt="Test image"
    className="drphubp asqlr8n"
    loading="lazy"
    src="placeholder.jpg"
    srcSet={null}
    style={
      Object {
        "background": "#fff",
        "color": "#fff",
        "filter": "blur(100px)",
        "transform": "scale3d(1.05,1.05,1)",
        "transition": "initial",
      }
    }
  />
</a>
`;

exports[`react customized options for normal image 1`] = `
<div
  className="a1y806g2"
>
  <div
    className="aund3wh"
    style={
      Object {
        "background": "#fff",
        "color": "#fff",
        "paddingBottom": "80%",
      }
    }
  />
  <img
    alt="Test image"
    className="drphubp asqlr8n"
    loading="lazy"
    src="image.jpg"
    srcSet={null}
    style={
      Object {
        "background": "#fff",
        "color": "#fff",
        "filter": "initial",
        "transform": "initial",
        "transition": "initial",
      }
    }
  />
</div>
`;

exports[`react parse valid input 1`] = `
<div
  className="a1y806g2"
>
  <div
    className="aund3wh"
    style={
      Object {
        "background": "transparent",
        "color": "transparent",
        "paddingBottom": "80%",
      }
    }
  />
  <img
    alt="Test image"
    className="drphubp asqlr8n"
    loading="lazy"
    src="image.jpg"
    srcSet={null}
    style={
      Object {
        "background": "transparent",
        "color": "transparent",
        "filter": "initial",
        "transform": "initial",
        "transition": "initial",
      }
    }
  />
</div>
`;
